kroxylicious-sample/src/test/java/io/kroxylicious/sample/SampleProduceRequestFilterTest.java

```diff
@@ -0,0 +1,157 @@
+    private static final String NO_TRANSFORM_VALUE = "this value will not be transformed";
+    private static final String POST_TRANSFORM_VALUE = "this is what the value will be transformed to";
+    private static final String CONFIG_FIND_VALUE = "from";
+    private static final String CONFIG_REPLACE_VALUE = "to";
+    private static final ApiMessageType API_MESSAGE_TYPE = ApiMessageType.PRODUCE;
+
+    @Mock
+    private KrpcFilterContext context;
+
+    @Captor
+    private ArgumentCaptor<ApiMessage> apiMessageCaptor = ArgumentCaptor.forClass(ApiMessage.class);
+
+    private FilterInvoker invoker;
+
+    @BeforeEach
+    public void beforeEach() {
+        buildContextMock();
+        SampleFilterConfig config = new SampleFilterConfig(CONFIG_FIND_VALUE, CONFIG_REPLACE_VALUE);
+        SampleProduceRequestFilter filter = new SampleProduceRequestFilter(config);
+        invoker = FilterInvokers.from(filter);
+    }
+
+    /**
+     * Unit Test: Checks that transformation is applied when request data contains configured value.
+     */
+    @Test
+    public void willTransformProduceRequestTest() {
+        var headerData = new RequestHeaderData();
+        var requestData = buildProduceRequestData(PRE_TRANSFORM_VALUE);
+        invoker.onRequest(ApiKeys.forId(API_MESSAGE_TYPE.apiKey()), API_VERSION, headerData, requestData, context);
+        verify(context).forwardRequest(any(), apiMessageCaptor.capture());
+        var unpackedRequest = unpackProduceRequestData((ProduceRequestData) apiMessageCaptor.getValue());
+        // We only put 1 record in, we should only get 1 record back, and
+        // We should see that the unpacked request value has changed from the input value, and
+        // We should see that the unpacked request value has been transformed to the correct value
+        assertThat(unpackedRequest)
+                .hasSize(1)
+                .first().asString()
+                .doesNotContain(PRE_TRANSFORM_VALUE)
+                .contains(POST_TRANSFORM_VALUE);
+    }
+
+    /**
+     * Unit Test: Checks that transformation is not applied when request data does not contain configured
+     * value.
+     */
+    @Test
+    public void wontTransformProduceRequestTest() {
+        var headerData = new RequestHeaderData();
+        var requestData = buildProduceRequestData(NO_TRANSFORM_VALUE);
+        invoker.onRequest(ApiKeys.forId(API_MESSAGE_TYPE.apiKey()), API_VERSION, headerData, requestData, context);
+        verify(context).forwardRequest(any(), apiMessageCaptor.capture());
+        var unpackedRequest = unpackProduceRequestData((ProduceRequestData) apiMessageCaptor.getValue());
+        // We only put 1 record in, we should only get 1 record back, and
+        // We should see that the unpacked request value has not changed from the input value
+        assertThat(unpackedRequest)
+                .hasSize(1)
+                .first().asString()
+                .contains(NO_TRANSFORM_VALUE);
```

Comment:
```suggestion
        assertThat(unpackedRequest)
                .containsExactly(NO_TRANSFORM_VALUE);
```

Reply:
Unfortunately this doesn't work. At some point (I managed to track it to prior to sending to the filter, but not any further than that) we're getting a bunch of seemingly random characters inserted at the start of our fetch data, which means using `containsExactly` produces the following failure:
```
[INFO] Running io.kroxylicious.sample.SampleFetchResponseFilterTest
[ERROR] Tests run: 2, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.008 s <<< FAILURE! -- in io.kroxylicious.sample.SampleFetchResponseFilterTest
[ERROR] io.kroxylicious.sample.SampleFetchResponseFilterTest.wontTransformFetchResponseTest -- Time elapsed: 0.003 s <<< FAILURE!
org.opentest4j.AssertionFailedError: 

Expecting actual:
  ["Z�����S�@������������������������������PDthis value will not be transformed"]
to contain exactly (and in same order):
  ["this value will not be transformed"]
but some elements were not found:
  ["this value will not be transformed"]
and others were not expected:
  ["Z�����S�@������������������������������PDthis value will not be transformed"]
```
Would love to have this doing exact matches, because that's just a better test, but I haven't managed to get ti to work. Worth noting that `SampleProduceRequestFilterTest` has the same problem.

Reply:
might be a nice convenience for Kroxylicious to provide as I've written stuff like this a couple of times recently, like a couple of static methods to go from `Record` to key and value `byte[]` (or String)

Reply:
The issues with retrieving record values have been fixed and it now uses `containsExactly` rather than the partial match `contains`.